### PR TITLE
Rewrite observer system

### PR DESCRIPTION
**Pull request by basine**
**In this PR**

**Note**: There are still a lot of TODOs  (search for TODO Sabine) in this PR. So this is still WIP. Some of them are open questions, feel free to add your comments to those questions.

Our observer system is currently very slow due to creating snapshots for almost every object in the object graph. This leads to long startup times and probably also high energy consumption. Instead of keeping snapshots of all objects, we are only creating snapshots of changed objects before the merge by getting the updated objects from the `NSManagedObjectContextDidSaveNotification` of the other context and creating snapshots for them on the other context.

For every ManagedObject we map the `keysPathsForValuesAffectedByValueForKey` of keys we want to observe in a `DependencyKeystore` object.

Every time an object changes on the context, we get the changedKeys, check for changes in dependent objects using the mapped keys in the `DependencyKeystore` and then merge all the changes we found with previous changes. When the contexts are merged or the uiContext saves, we calculate changeInfos and then post them using NSNotificationCenter.

There are serval additional objects:
`VoicechannelObserverCenter`
`ConversationListObserverCenter`
`MessageWindowObserverCenter`
`SearchUserObserverCenter`
which are forwarded the changes calculated during the save and update their lists or snapshots accordingly.

**NotificationDispatcher**
* Creates Snapshots of objects before merge (via SnapshotCenter)
* Processes NSManagedObjectsDidChangeNotification getting either changedValues() or using the snapshot to retrieve changes if the object is faulted
* Calculates changes in other objects due to insert / deletes / updates
    * Uses reverse map of `keyPathsForValueAffectingValueForKey`
    * Checks if related objects depend on changes using “manual” relationship mappings in `SideEffectSources`:
E.g. a change of a user’s name has an effect on it’s conversations’ displayNames, therefore it would add the conversations and all keys that are affected by “connection.to.name” or “otherActiveParticipants.name”
Since displayName is a computed property, this change would not be noticed just inspecting changedValues or the snapshot
* When contextDidSave or contextDidMerge is called:
    * It creates ObjectChangeInfos from the collected changes 
    * It fires an NSNotification whose userInfo contains these changeInfos (subscribers subscribe to the NotificationCenter directly which maps the callback in its callback block)
    * It forwards those changeInfos to other changeInfoConsumers who were added
* When the app enters the background
    * It stops processing and forwarding changes
    * It resets all current changes and clears all snapshots
    * It notifies all changeInfoConsumers that it went to the background
* When the app enters the foreground:
    * It starts forwarding the changes again
    * It notifies all changeInfoConsumers that it went to the foreground
* In addition:
    * Calculates changes due to userImage changes or displayName changes	
    * It notifies about new unread / unsent messages
    * It notifies about non-CoreData changes via a notification call

**ConversationListObserverCenter**
* When a list is created, the list is added to the center (through a call in it’s init) and a snapshot created
* The snapshots are updated when a conversation is inserted / deleted / updated	
* When the app enters the background the list snapshots are reset
* When the app enters the foreground:
    * The UI calls a method to recreate all conversation lists which calls the center
    * The list snapshots are recreated

**MessageWindowObserverCenter**
* When a window is created, a snapshot is created - there can only be one window at a time therefore there will only be one snapshot at a time
* The snapshots are updated 
    * when a message is inserted / deleted / updated
    * when a user is updated whose message are contained in window
* When the app enters the background we do nothing
* When the app enters the foreground the window is recalculated and the change notification sent

**SearchuserObserverCenter**
* When the searchDirectory returns new searchUsers, snapshots for those searchusers are created when calling `searchDirectoryDidUpdate`
* When the searchDirectory is tornDown, the observer center snapshots are reset
* The snapshots are updated when:
    * The searchUser’s user changes and we receive a UserChangeInfo
    * The method `notifyUpdatedSearchUser` is called when the searchUser’s image changes
* When the app enters the background we do nothing
* When the app enters the foreground we do nothing

### CODE REVIEW

File: Source/Notifications/DependencyKeyStore.swift [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+
+/// Maps the observable keys to affectedKeys and vice versa
+/// You should create this only once
+class DependencyKeyStore {
+    
+    /// Keys that are needed to create a changeInfo
+    let observableKeys : [String: Set<String>]
+    
+    /// All keys that will create a userChangeInfo
```
Comment (daehn) [inline Source/Notifications/DependencyKeyStore.swift]: I think the `userChangeInfo` in the comment should be a `changeInfo`

**Comment by typfel** [comment]
Lot's of code to look into :) First have some questions:

1. From the initial meeting I went away with the impression that would not do snapshotting at all anymore and instead fire a `allKeys`-changed-notification. What makes it necessary to still do snapshots?

2. Looking at the`NotificationDispatcher` I see that it's coupled to the entities. I was hoping we could have core entity agnostic change dispatcher. The reason for this is that I want to have the possibility to implement an observer also at a higher level like I currently do for the `WireCallCenterV2` which moved to sync-engine. This ties into my wish of having a more modular observation system.

**Comment by basine** [comment]
@typfel 
Regarding your questions: 
>From the initial meeting I went away with the impression that we would not do snapshotting at all >anymore and instead fire a allKeys-changed-notification. What makes it necessary to still do >snapshots?

Basically every time the contexts merge objects that were updated in the sync Context will get refreshed and faulted in the uiMOC. Therefore we still need the snapshots - we only snapshot those elements that are updated in the synMOC before the merge, so it's a lot fewer. Otherwise you would get an allKeys changed for a lot of cases.

>Looking at theNotificationDispatcher I see that it's coupled to the entities. I was hoping we could ?>have core entity agnostic change dispatcher. The reason for this is that I want to have the >possibility to implement an observer also at a higher level like I currently do for the >WireCallCenterV2 which moved to sync-engine. This ties into my wish of having a more modular >observation system.

To do this you would have to specify further what you need exactly for this. Since we want to observe changes in ManagedObjects as well as changes that are a result of other object's changes it is inevitable that the dispatcher will be CoreData heavy. Once this thing works, we can abstract it more and maybe fit it to your needs, but before that I don't think it makes sense to abstract it further.

**Comment by typfel** [comment]
NotificationDispatcher currently (as I understood it)
- does a mapping from core data change to changeInfos (UserChangeInfo, ...) and posts them in notifications
- has a references to all centers and tell them to also post their changes

I thought `NotificationDispatcher` could fire a entity agnostic notification which would be a more reliable version of the core data change notifications and also us to track non core data objects.

Something along the lines of:

ChangeNotification {
	updated: [ObjectChange]
	added: ..
	deleted ..
}

ObjectChange {
	object: Any // ZMUser for example
	changedKeys : Set<String>
}

It would then be the responsibillity of another class, build on top of this more reliable change notification, to produce `UserChangeInfo` notifications. The conversation list center would then also listen to these notifications and would only be coupled to the `NotificationDispatcher` via its notifications.

**Comment by basine** [comment]
@typfel It does not really post `ObjectChangeInfo` objects, instead it posts an NSNotification. When an observer registers it registers under the hood in NSNotificationCenter. The ChangeInfo is extracted from the NSNotification and in the observer callBack block the respective observer protocol function called. This already makes it more flexible to include non coreData objects. All you need to do is specify a static classIdentifier for the class and a notificationName for this classIdentifier. At the moment it's using the ZMManagedObjects entityName, but this can be abstracted (there is a TODO for that) in a protocol so that any class that implements the protocol can dispatch notifications using the dispatcher.

File: Source/Notifications/DependencyKeyStore.swift [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        self.affectingKeys = affecting
+        self.allKeys = all
+        self.effectedKeys = affectingInverse
+    }
+    
+    /// When adding objects that are to be observed, add keys that are supposed to be reported on in here
+    private static func setupObservableKeys(classIdentifier: String) -> Set<String> {
+        switch classIdentifier {
+        case ZMConversation.entityName():
+            return Set(arrayLiteral: "messages", "lastModifiedDate", "isArchived", "conversationListIndicator", "voiceChannelState", "activeFlowParticipants", "callParticipants", "isSilenced", "securityLevel", "otherActiveVideoCallParticipants", "displayName", "estimatedUnreadCount", "clearedTimeStamp", "otherActiveParticipants", "isSelfAnActiveMember", "relatedConnectionState")
```
Comment (daehn) [inline Source/Notifications/DependencyKeyStore.swift]: Would bench to move those into the observed classes. Or to have a protocol `ObservableType` with a method `observableKeys` that the objects need to conform to. Also I did not try it yet myself, but generally in Swift it should be possible to use `#keyPath(ZMConversation.messages)` which will return a `String`.

File: Source/Notifications/VoicechannelObserverCenter.swift [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+                                                     object: conversation,
+                                                     userInfo: ["changeInfo": stateChange]))
+    }
+}
+
+
+
+public extension NSOrderedSet {
+    
+    public func subtracting(orderedSet: NSOrderedSet) -> NSOrderedSet {
```
Comment (daehn) [inline Source/Notifications/VoicechannelObserverCenter.swift]: Could be moved into `wire-iso-utilies` to make them available in other frameworks as well.

File: Source/Notifications/ZMNotifications+Internal.h [hunk 4]
```diff
@@ -31,31 +31,31 @@
...
-
-@interface ZMSearchUser (InternalObservers)
-
-+ (id<ZMUserObserverOpaqueToken>)addUserObserver:(id<ZMUserObserver>)observer forUsers:(NSArray *)users managedObjectContext:(NSManagedObjectContext *)managedObjectContext;
-
-@end
-
+//
+//@interface ZMMessageNotification (Internal)
+//
```
Comment (daehn) [inline Source/Notifications/ZMNotifications+Internal.h]: Can be removed?

File: Source/Notifications/ZMNotifications.m [hunk 5]
```diff
@@ -113,172 +113,172 @@ - (NSUInteger)hash
...
-+ (id<ZMUserObserverOpaqueToken>)addUserObserver:(id<ZMUserObserver>)observer forUsers:(NSArray *)users inUserSession:(id<ZMManagedObjectContextProvider>)userSession;
-{
-    return [ZMUser addUserObserver:observer forUsers:users managedObjectContext:userSession.managedObjectContext];
-}
-
-@end
-
+//
+//
+//@implementation ZMConversationMessageWindow (Notifications)
```
Comment (daehn) [inline Source/Notifications/ZMNotifications.m]: Can be removed?

File: Source/ConversationList/ZMConversationList.m [hunk 6]
```diff
@@ -87,6 +87,12 @@ - (NSManagedObjectContext *)managedObjectContext
     return self.moc;
 }
 
+- (void)recreateWithAllConversation:(NSArray *)conversations
```
Comment (mikeger) [inline Source/ConversationList/ZMConversationList.m]: Typo `recreateWithAllConversation`->`recreateWithAllConversations`

File: Source/ManagedObjectContext/NSManagedObjectContext+zmessaging.m [hunk 7]
```diff
@@ -180,7 +180,8 @@ + (instancetype)createUserInterfaceContextWithStoreAtURL:(NSURL *)storeURL
     [SharedUserInterfaceContext initialiseSessionAndSelfUser];
     
     SharedUserInterfaceContext.mergePolicy = [[ZMSyncMergePolicy alloc] initWithMergeType:NSRollbackMergePolicyType];
-    (void)SharedUserInterfaceContext.globalManagedObjectContextObserver;
+// TODO Sabine: it would probably be better to initialize the dispatcher in the syncstrategy instead? Or do we need it for other purposes?
+    //    (void)SharedUserInterfaceContext.globalManagedObjectContextObserver;
```
Comment (mikeger) [inline Source/ManagedObjectContext/NSManagedObjectContext+zmessaging.m]: TODO left

File: Source/Model/Message/ZMClientMessage+Encryption.swift [hunk 8]
```diff
@@ -261,3 +263,26 @@ extension ZMGenericMessage {
         return externalGenericMessage.encryptedMessagePayloadData(conversation, externalData: encryptedDataWithKeys.data)
     }
 }
+
+// MARK: - Session identifier {
+extension UserClient {
```
Comment (mikeger) [inline Source/Model/Message/ZMClientMessage+Encryption.swift]: So this PR includes Marco's session identifier changes?

File: Source/Model/User/ZMUserDisplayNameGenerator.m [hunk 9]
```diff
@@ -162,6 +162,7 @@ - (NSSet *)updateDisplayNameGeneratorWithChanges:(NSNotification *)note;
     }
     
     // loop through updated. If the user name changed then replace those users in the updatedSet
+    // TODO Sabine: This does not make sense, user is a reference and should change values as the MO changes
```
Comment (mikeger) [inline Source/Model/User/ZMUserDisplayNameGenerator.m]: /On behalf of TODO bot/ TODO left!

File: Source/Notifications/ConversationListObserverCenter.swift [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+    static let ZMConversationListDidChange = Notification.Name("ZMConversationListDidChangeNotification")
+}
+
+let ConversationListObserverCenterKey = "ConversationListObserverCenterKey"
+
+extension NSManagedObjectContext {
+    
+    public var conversationListObserverCenter : ConversationListObserverCenter? {
+        guard zm_isUserInterfaceContext else {
+            zmLog.warn("ConversationListObserver does not exist in syncMOC")
```
Comment (mikeger) [inline Source/Notifications/ConversationListObserverCenter.swift]: Should be an assert?

File: Source/Notifications/ConversationListObserverCenter.swift [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+    func tearDown() {
+        if isTornDown { return }
+        isTornDown = true
+        
+        listSnapshots.values.forEach{$0.tearDown()}
+        listSnapshots = [:]
+    }
+    
+    public func applicationDidEnterBackground() {
+        // TODO Sabine: clear snapshots?
```
Comment (mikeger) [inline Source/Notifications/ConversationListObserverCenter.swift]: /on behalf of TODO bot/

File: Source/Notifications/ConversationListObserverCenter.swift [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+        listSnapshots = [:]
+    }
+    
+    public func applicationWillEnterForeground() {
+        // do nothing?
```
Comment (mikeger) [inline Source/Notifications/ConversationListObserverCenter.swift]: /on behalf of TODO bot/

File: Source/Notifications/MessageWindowObserverCenter.swift [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+            if let messageChanges = $0 as? [MessageChangeInfo] {
+                messageChanges.forEach{snapshot.messageDidChange($0)}
+            }
+        }
+        
+        snapshot.fireNotifications()
+    }
+    
+    public func applicationDidEnterBackground() {
+        // clear snapshot?
```
Comment (mikeger) [inline Source/Notifications/MessageWindowObserverCenter.swift]: /on behalf of TODO bot/

File: Source/Notifications/MessageWindowObserverCenter.swift [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+    }
+    
+    public func applicationWillEnterForeground() {
+        // do nothing?
```
Comment (mikeger) [inline Source/Notifications/MessageWindowObserverCenter.swift]: /on behalf of TODO bot/

File: Source/Notifications/NotificationDispatcher.swift [hunk 15]
```diff
@@ -0,0 +1,561 @@
...
+        guard let object = note.object as? ZMManagedObject,
+              let changedKeys = (note.userInfo as? [String : [String]])?["changedKeys"]
+        else { return }
+        
+        let classIdentifier = type(of: object).classIdentifier
+        let change = Changes(changedKeys: Set(changedKeys))
+
+        let objectAndChangedKeys = [object: change]
+        allChanges[classIdentifier] = allChanges[classIdentifier]?.merged(with: objectAndChangedKeys) ?? objectAndChangedKeys
+        // TODO Sabine: make sure that save is always called
```
Comment (mikeger) [inline Source/Notifications/NotificationDispatcher.swift]: /on behalf of TODO bot/

File: Source/Notifications/NotificationDispatcher.swift [hunk 16]
```diff
@@ -0,0 +1,561 @@
...
+    @objc func contextDidSave(_ note: Notification){
+        guard forwardChanges else { return }
+        forwardChangesToConversationListObserver(note: note)
+        fireAllNotifications()
+    }
+    
+    /// This will be called if a change to an object does not cause a change in Core Data, e.g. downloading the asset and adding it to the cache
+    @objc func nonCoreDataChange(_ note: Notification){
+        guard forwardChanges else { return }
+        // TODO Sabine: add tests for this!
```
Comment (mikeger) [inline Source/Notifications/NotificationDispatcher.swift]: /on behalf of TODO bot/

File: Source/Notifications/ObjectObserverTokens/ConversationChangeInfo.swift [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+        let changeInfo = ConversationChangeInfo(object: conversation)
+        changeInfo.changedKeysAndOldValues = changes.originalChanges
+        changeInfo.changedKeys = changes.changedKeys
+        return changeInfo
+    }
+}
+
+
+
+//@objc public protocol ZMConversationObserverOpaqueToken :NSObjectProtocol {}
```
Comment (mikeger) [inline Source/Notifications/ObjectObserverTokens/ConversationChangeInfo.swift]: Commented code

File: Source/Notifications/ObjectObserverTokens/ConversationListChangeInfo.swift [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+    public var conversationList : ZMConversationList { return self.observedObject as! ZMConversationList }
+    
+    init(setChangeInfo: SetChangeInfo) {
+        super.init(observedObject: setChangeInfo.observedObject, changeSet: setChangeInfo.changeSet)
+    }
+}
+
+
+
+//@objc public protocol ZMConversationListObserverOpaqueToken : NSObjectProtocol {}
```
Comment (mikeger) [inline Source/Notifications/ObjectObserverTokens/ConversationListChangeInfo.swift]: Commented code

File: Source/Notifications/ObjectObserverTokens/MessageChangeInfo.swift [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+        return changedKeysAndOldValues[MessageChangeInfo.ReactionChangeInfoKey] as? [ReactionChangeInfo] ?? []
+    }
+    
+    public let message : ZMMessage
+    
+}
+
+
+
+//@objc public protocol ZMMessageObserverOpaqueToken : NSObjectProtocol {}
```
Comment (mikeger) [inline Source/Notifications/ObjectObserverTokens/MessageChangeInfo.swift]: Commented code

File: Source/Notifications/ObjectObserverTokens/MessageWindowChangeInfo.swift [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+    convenience init(windowWithMissingMessagesChanged window: ZMConversationMessageWindow, isFetching: Bool) {
+        self.init(setChangeInfo: SetChangeInfo(observedObject: window))
+        self.isFetchingMessages = isFetching
+        self.isFetchingMessagesChanged = true
+        
+    }
+}
+
+
+//@objc public protocol ZMConversationMessageWindowObserverOpaqueToken : NSObjectProtocol {}
```
Comment (mikeger) [inline Source/Notifications/ObjectObserverTokens/MessageWindowChangeInfo.swift]: Commented code

File: Source/Notifications/ObjectObserverTokens/NewUnreadMessageChangeInfos.swift [hunk 21]
```diff
@@ -0,0 +1,162 @@
...
+    }
+    
+    public var messages : [ZMConversationMessage] {
+        return object as? [ZMConversationMessage] ?? []
+    }
+    
+}
+
+
+//@objc public protocol ZMNewUnreadMessageObserverOpaqueToken : NSObjectProtocol {}
```
Comment (mikeger) [inline Source/Notifications/ObjectObserverTokens/NewUnreadMessageChangeInfos.swift]: Commented code

File: Source/Notifications/ObjectObserverTokens/NewUnreadMessageChangeInfos.swift [hunk 22]
```diff
@@ -0,0 +1,162 @@
...
+        self.init(object: messages as NSObject)
+    }
+    
+    public var messages : [ZMConversationMessage] {
+        return object as? [ZMConversationMessage] ?? []
+    }
+}
+
+
+//@objc public protocol ZMNewUnreadKnockMessageObserverOpaqueToken : NSObjectProtocol {}
```
Comment (mikeger) [inline Source/Notifications/ObjectObserverTokens/NewUnreadMessageChangeInfos.swift]: Code commented

File: Source/Notifications/ObjectObserverTokens/UserChangeInfo.swift [hunk 23]
```diff
@@ -0,0 +1,212 @@
...
+
+    open let user: ZMBareUser
+    open var userClientChangeInfos : [UserClientChangeInfo] {
+        return changedKeysAndOldValues[UserChangeInfo.UserClientChangeInfoKey] as? [UserClientChangeInfo] ?? []
+    }
+}
+
+
+
+//@objc public protocol ZMUserObserverOpaqueToken : NSObjectProtocol {}
```
Comment (mikeger) [inline Source/Notifications/ObjectObserverTokens/UserChangeInfo.swift]: Code is commented

File: Source/Notifications/ObjectObserverTokens/UserClientChangeInfo.swift [hunk 24]
```diff
@@ -0,0 +1,110 @@
...
+        let changeInfo = UserClientChangeInfo(object: client)
+        changeInfo.changedKeysAndOldValues = changes.originalChanges
+        changeInfo.changedKeys = changes.changedKeys
+        return changeInfo
+    }
+    
+}
+
+
+//@objc public protocol UserClientObserverOpaqueToken : NSObjectProtocol {}
```
Comment (mikeger) [inline Source/Notifications/ObjectObserverTokens/UserClientChangeInfo.swift]: The code is commented

File: Source/Notifications/ObjectObserverTokens/VoiceChannelChangeInfos.swift [hunk 25]
```diff
@@ -0,0 +1,151 @@
...
+    }
+    public var voiceChannel : ZMVoiceChannel? { return (object as? ZMConversation)?.voiceChannel }
+    
+    public override var description: String {
+        return "Call state changed from \(previousState) to \(currentState)"
+    }
+}
+
+
+//@objc public protocol ZMVoiceChannelStateObserverOpaqueToken : NSObjectProtocol {}
```
Comment (mikeger) [inline Source/Notifications/ObjectObserverTokens/VoiceChannelChangeInfos.swift]: Well, you know, I guess it's all the same code commented, but I have to note so you would see where to remove.

File: Source/Notifications/ObjectObserverTokens/VoiceChannelChangeInfos.swift [hunk 26]
```diff
@@ -0,0 +1,151 @@
...
+    }
+    
+    let conversation : ZMConversation
+    public var voiceChannel : ZMVoiceChannel { return conversation.voiceChannel }
+    public var otherActiveVideoCallParticipantsChanged : Bool = false
+    
+}
+
+
+//@objc public protocol ZMVoiceChannelParticipantsObserverOpaqueToken : NSObjectProtocol {}
```
Comment (mikeger) [inline Source/Notifications/ObjectObserverTokens/VoiceChannelChangeInfos.swift]: And here

File: Source/VoiceChannel/ZMVoiceChannel.m [hunk 27]
```diff
@@ -151,7 +151,8 @@ - (instancetype)initWithConversation:(ZMConversation *)conversation;
         _conversation = conversation;
         NSManagedObjectContext *context = self.conversation.managedObjectContext;
         if (context.zm_isUserInterfaceContext) {
-            self.callCenter = context.globalManagedObjectContextObserver.callCenter;
+            // TODO Sabine: replicate with dispatcher?
+//            self.callCenter = context.globalManagedObjectContextObserver.callCenter;
```
Comment (mikeger) [inline Source/VoiceChannel/ZMVoiceChannel.m]: /on behalf of TODO bot/

File: Tests/Source/Model/Observer/ConversationListObserverTests.swift [hunk 28]
```diff
@@ -701,57 +679,54 @@ class GlobalConversationObserverTests : ZMBaseManagedObjectTest {
...
-        self.uiMOC.globalManagedObjectContextObserver.notifyUpdatedCallState(Set(arrayLiteral: conversation1, conversation2, conversation3), notifyDirectly: true)
         
         // then
-        XCTAssertEqual(testObserver.changes.count, 1);
-        
-        self.uiMOC.globalManagedObjectContextObserver.removeGlobalVoiceChannelStateObserverForToken(token)
+        XCTAssertEqual(testObserver.changes.count, 0)
     }
+    
+//    func testThatItNotifiesGlobalVoiceChannelObserversWhenAVoiceChannelStateChanges()
```
Comment (mikeger) [inline Tests/Source/Model/Observer/ConversationListObserverTests.swift]: Commented code -- I guess you are waiting for Jacob's changes

File: Tests/Source/Model/Observer/ManagedObjectContextObserverTests.swift [hunk 29]
```diff
@@ -18,316 +18,317 @@
...
-            XCTFail("No note")
-        }
-        
-        // after
-        ZMMessageNotification.removeMessageObserver(for: messageObserver)
-        
-    }
-    
-}
+// TODO Sabine: Remove when not needed anymore
```
Comment (mikeger) [inline Tests/Source/Model/Observer/ManagedObjectContextObserverTests.swift]: Should be removed?

File: Tests/Source/Model/Observer/MessageWindowObserverTests.swift [hunk 30]
```diff
@@ -0,0 +1,542 @@
...
+            let token = MessageWindowChangeInfo.add(observer: self.windowObserver, for: window)
+            
+            self.startMeasuring()
+            for _ in 1...count {
+                let message = ZMClientMessage.insertNewObject(in: self.uiMOC)
+                message.visibleInConversation = window.conversation
+                self.uiMOC.saveOrRollback()
+            }
+            self.stopMeasuring()
+//            XCTAssertEqual(self.windowObserver.notifications.count, count)
```
Comment (mikeger) [inline Tests/Source/Model/Observer/MessageWindowObserverTests.swift]: Commented code

File: Source/Model/User/ZMUserDisplayNameGenerator.m (hunk 9 above)
Comment (basine) [inline Source/Model/User/ZMUserDisplayNameGenerator.m]: There are other TODOs left - this is still WIP until Marco's and Jacob's changes have been merged

File: Tests/Source/Model/Observer/ObjectSnapshotTests.swift [hunk 31]
```diff
@@ -21,255 +21,256 @@ import Foundation
...
-                (data : NSMutableData) in data.append(Data(bytes: [1,2,3,4]))
-        }))
-        XCTAssertTrue(checkThatItReturnsANewSnapshotIfThereIsANewObject("string",
-            mutator: {
-                (string : NSMutableString) in string.append("bla")
-        }))
-    }
-}
-
+// TODO Sabine: Remove when not needed
```
Comment (mikeger) [inline Tests/Source/Model/Observer/ObjectSnapshotTests.swift]: Should be removed?

File: Source/Model/Message/ZMClientMessage+Encryption.swift (hunk 8 above)
Comment (basine) [inline Source/Model/Message/ZMClientMessage+Encryption.swift]: Yes, I merged Marco's changes too early.

**Pull request merged**